### PR TITLE
Learning rate schedule is now a config class

### PR DESCRIPTION
**Pull request by hyenal**
To avoid to over parametrise the learning rate schedule class I propose to make them serializable objects. This is also related to https://github.com/martinsbruveris/tensorflow-image-models/discussions/38

### CODE REVIEW

**Comment by martinsbruveris** [comment]
One comment about code style: please run `make style` followed by `make quality` on your code.
- `make style` applies black formatting and import sorting
- `make quality` checks black formatting, import sorting and also runs `flake8`

File: tfimm/train/optimizers/lr_schedules.py [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+from dataclasses import dataclass
+from abc import ABC, abstractclassmethod
+
+import tensorflow as tf
+from tensorflow.keras.optimizers.schedules import LearningRateSchedule
+
+from ..registry import cfg_serializable, get_class
+
+
+class BaseLrSchedule(ABC):
```
Comment (martinsbruveris) [inline tfimm/train/optimizers/lr_schedules.py:11]: Can I suggest using `LR` instead of `Lr` in the code. Given that it is an abbreviation I would suggest capitalising both. I am also using, e.g., `MLP` elsewhere in the code.

File: tfimm/train/__init__.py [hunk 2]
```diff
@@ -16,7 +16,17 @@
...
+from .optimizers.optimizer import OptimizerConfig, OptimizerFactory  # noqa: F401
+from .optimizers.lr_schedules import (
+    LrConstConfig,
+    ConstSchedule,
+    LrMultiStepsConfig,
+    MultiStepsSchedule,
+    LrCosineDecayConfig,
+    CosineDecaySchedule,
+    LrExpDecayConfig,
+    ExponentialDecaySchedule,
```
Comment (martinsbruveris) [inline tfimm/train/__init__.py:28]: Regarding naming, can I suggest synchronising the names of the config class and the schedule, i.e.,
- `LRConstConfig`, `LRConstFactory`
- `LRMultiStepsConfig`, `LRMultiStepsFactory`
- `LRCosineDecayConfig`, `LRCosineDecayFactory`
- `LRExpDecayConfig`, `LRExpDecayFactory`
The reason for calling the Factory rather than Schedule is the same as for `OptimizerFactory`: they are classes to create LR schedules.

**Comment by martinsbruveris** [comment]
Another small suggestion: rename `lr_schedules.py` to simply `schedules.py`. To keep names short.

Looking good! Thanks for the PR.

**Comment by hyenal** [comment]
Thanks :) It now pass the test but I still have to modify the examples to modify the schedule there :)

File: tests/test_architectures.py [hunk 3]
```diff
@@ -76,7 +76,9 @@ def test_load_timm_model(model_name: str):
     assert (np.max(np.abs(tf_res - pt_res))) / (np.max(np.abs(pt_res)) + 1e-6) < 1e-3
 
 
-@pytest.mark.parametrize("model_name", list_models(module="convnext", exclude_filters=EXCLUDE_FILTERS))
+@pytest.mark.parametrize(
+    "model_name", list_models(module="convnext", exclude_filters=EXCLUDE_FILTERS)
```
Comment (martinsbruveris) [inline tests/test_architectures.py:80]: Could you please change this line to
```
@pytest.mark.parametrize("model_name", list_models(exclude_filters=EXCLUDE_FILTERS))
```
I forgot that I added this, when I was adding the ConvNeXt models....

**Comment by martinsbruveris** [comment]
Looks good! Will merge once unit tests pass.

**Comment by hyenal** [comment]
Test linked to this MR are passing but some errors not linked to it have appeared. How should I go about it ?
